File: uat/custom-components/client-java-sdk/src/main/java/com/aws/greengrass/testing/mqtt5/client/GRPCLib.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.aws.greengrass.testing.mqtt5.client;

import com.aws.greengrass.testing.mqtt5.client.exceptions.GRPCException;
import lombok.NonNull;

/**
 * Interface of gRPC library.
 */
public interface GRPCLib {
    /**
     * Creates and establishes bidirectional link with the client control.
     *
     * @param agentId the id of agent to identify control channel by control
     * @param hosts the array of host name or IP address of gRPC server to connect to
     * @param port the TCP port to connect to
     * @return connection handler
     * @throws GRPCException on errors
     */
    GRPCLink makeLink(@NonNull String agentId, @NonNull String[] hosts, int port) throws GRPCException;
}

File: uat/custom-components/client-mosquitto-c/src/GRPCLib.cpp
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <grpcpp/health_check_service_interface.h>                      /* EnableDefaultHealthCheckService() */
#include <grpcpp/ext/proto_server_reflection_plugin.h>                  /* InitProtoReflectionServerBuilderPlugin() */

#include "GRPCLib.h"                    /* self header */
#include "GRPCLink.h"                   /* class GRPCLink */
#include "logger.h"                     /* logd() */


GRPCLib::GRPCLib() {
    logd("Initialize gRPC library\n");

    grpc::EnableDefaultHealthCheckService(true);
    grpc::reflection::InitProtoReflectionServerBuilderPlugin();
}

GRPCLib::~GRPCLib() {
    logd("Shutdown gRPC library\n");
}

GRPCLink * GRPCLib::makeLink(const std::string & agent_id, const std::list<std::string> & hosts, unsigned short port) {
    return new GRPCLink(agent_id, hosts, port);
}
